pkg/gatewayserver/io/io.go

```diff
@@ -201,26 +201,48 @@ var errBufferFull = errors.DefineInternal("buffer_full", "buffer is full")
 // Interval between emitting consecutive gs.up.repeat events for the same gateway connection.
 const consecutiveRepeatUpEventsInterval = time.Minute
 
+// ManualClockSynchronization contains the clock synchronization
+// timestamps provided by a frontend for manual synchronization.
+type ManualClockSynchronization struct {
+	Timestamp        uint32
+	ServerTime       time.Time
+	GatewayTime      *time.Time
+	ConcentratorTime scheduling.ConcentratorTime
+}
+
 // HandleUp updates the uplink stats and sends the message to the upstream channel.
-func (c *Connection) HandleUp(up *ttnpb.UplinkMessage) error {
+func (c *Connection) HandleUp(up *ttnpb.UplinkMessage, manualSync *ManualClockSynchronization) error {
 	if c.discardRepeatedUplink(up) {
 		return nil
 	}
 
 	var ct scheduling.ConcentratorTime
-	if up.Settings.Time != nil {
+	switch {
+	case manualSync != nil:
+		ct = c.scheduler.SyncWithGatewayConcentrator(manualSync.Timestamp, manualSync.ServerTime, manualSync.GatewayTime, manualSync.ConcentratorTime)
+		log.FromContext(c.ctx).WithFields(log.Fields(
+			"timestamp", manualSync.Timestamp,
+			"concentrator_time", manualSync.ConcentratorTime,
+			"server_time", manualSync.ServerTime,
+			"gateway_time", manualSync.GatewayTime,
+		)).Info("Gateway clocks have been synchronized by the frontend")
```

Comment:
So then `pkg/gatewayserver/io.FrontendClockSynchronization` sounds like a great name to me